### PR TITLE
IGNITE-23202 Fix server node fail when nodeFilter class unknown

### PR DESCRIPTION
**Pull request by nizhikov**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

**Comment by Vladsz83** [comment]
The test `testClashingNameDifferentConfig()` fails.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/service/IgniteServiceDeployUnknownClassTest.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import org.apache.ignite.services.ServiceDeploymentException;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.junit.Test;
+
+import static org.apache.ignite.testframework.GridTestUtils.assertThrowsWithCause;
+
+/**
+ * Tests that not all nodes in cluster need user's service definition (only nodes according to filter).
+ */
+public class IgniteServiceDeployUnknownClassTest extends GridCommonAbstractTest {
```
Comment (Vladsz83) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/service/IgniteServiceDeployUnknownClassTest.java:36]: Not sure that we need a dedicated test class for just one test. There are similar tests. Maybe `IgniteServiceDeploymentClassLoadingDefaultMarshallerTest`?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java [hunk 2]
```diff
@@ -1764,11 +1781,21 @@ else if (req instanceof ServiceUndeploymentRequest)
...
                         err = checkPermissions(((ServiceDeploymentRequest)req).configuration().getName(), SERVICE_DEPLOY);
 
+                    if (err == null) {
+                        try {
+                            unmarshalNodeFilterIfNeeded(cfg);
+                        }
+                        catch (IgniteCheckedException e) {
+                            err = new IgniteCheckedException("Failed to deploy service, " +
+                                "unable to unmarshal node filter, cfg=" + cfg, e);
+                        }
```
Comment (Vladsz83) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java:1796]: Suggestion: "node filter" -> "service node filter". A bit better. We have cache node filter in Ignite.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/service/IgniteServiceDeploymentClassLoadingDefaultMarshallerTest.java [hunk 3]
```diff
@@ -152,6 +163,29 @@ public void testServiceDeployment3() throws Exception {
...
+                .newInstance(cli.context().localNodeId()))
+            .setTotalCount(1);
+
+        assertThrowsWithCause(() -> cli.services().deploy(svcCfg), ServiceDeploymentException.class);
+
+        assertTrue(cli.services().serviceDescriptors().isEmpty());
+        assertTrue(srv.services().serviceDescriptors().isEmpty());
+
+        // Check node alive.
+        srv.createCache(DEFAULT_CACHE_NAME).put(1, 1);
```
Comment (Vladsz83) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/service/IgniteServiceDeploymentClassLoadingDefaultMarshallerTest.java:186]: Suggestion: we might check the client too: `cli.cache(DEFAULT_CACHE_NAME).put(2, 2);`

Comment (NSAmelchev) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/service/IgniteServiceDeploymentClassLoadingDefaultMarshallerTest.java:186]: Fixed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/service/IgniteServiceDeploymentClassLoadingDefaultMarshallerTest.java [hunk 4]
```diff
@@ -152,6 +163,29 @@ public void testServiceDeployment3() throws Exception {
...
+                .getConstructor(UUID.class)
+                .newInstance(cli.context().localNodeId()))
+            .setTotalCount(1);
+
+        assertThrowsWithCause(() -> cli.services().deploy(svcCfg), ServiceDeploymentException.class);
+
+        assertTrue(cli.services().serviceDescriptors().isEmpty());
+        assertTrue(srv.services().serviceDescriptors().isEmpty());
+
+        // Check node alive.
```
[the remark this answers is not in the document]
Comment (NSAmelchev) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/service/IgniteServiceDeploymentClassLoadingDefaultMarshallerTest.java]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java (hunk 2 above)
Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java:1796]: Fixed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/service/IgniteServiceDeployUnknownClassTest.java (hunk 1 above)
Comment (NSAmelchev) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/service/IgniteServiceDeployUnknownClassTest.java]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java [hunk 5]
```diff
@@ -1134,7 +1136,21 @@ public <T> Collection<T> services(String name) {
     Map<UUID, Integer> reassign(@NotNull IgniteUuid srvcId, @NotNull ServiceConfiguration cfg,
         @NotNull AffinityTopologyVersion topVer,
         @Nullable TreeMap<UUID, Integer> oldTop) throws IgniteCheckedException {
-        Object nodeFilter = cfg.getNodeFilter();
+        IgnitePredicate<ClusterNode> nodeFilter = cfg.getNodeFilter();
+
+        if (cfg instanceof LazyServiceConfiguration) {
+            LazyServiceConfiguration srvcCfg = (LazyServiceConfiguration)cfg;
+
+            if (nodeFilter == null && srvcCfg.nodeFilterBytes() != null) {
```
[the remark this answers is not in the document]
Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java [hunk 6]
```diff
@@ -1764,11 +1781,21 @@ else if (req instanceof ServiceUndeploymentRequest)
                         "exists : [" + "srvcId" + reqSrvcId + ", srvcTop=" + oldDesc.topologySnapshot() + ']');
                 }
                 else {
-                    ServiceConfiguration cfg = ((ServiceDeploymentRequest)req).configuration();
+                    LazyServiceConfiguration cfg = ((ServiceDeploymentRequest)req).configuration();
 
                     if (ctx.security().enabled())
                         err = checkPermissions(((ServiceDeploymentRequest)req).configuration().getName(), SERVICE_DEPLOY);
 
+                    if (err == null) {
```
Comment (Vladsz83) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java:1789]: Under `if (err == null)` below?

```
if (err == null) {
    oldDesc = lookupInRegisteredServices(cfg.getName());

    if (oldDesc == null) {
```

File: modules/core/src/test/java/org/apache/ignite/internal/processors/service/IgniteServiceDeploymentClassLoadingDefaultMarshallerTest.java [hunk 7]
```diff
@@ -152,6 +163,30 @@ public void testServiceDeployment3() throws Exception {
         startClientGrid(CLIENT_NODE_WITH_EXT_CLASS_LOADER).services().deploy(serviceConfig());
     }
 
+    /** @throws Exception If failed. */
+    @Test
+    public void testFailWhenNodeFilterClassNotFound() throws Exception {
```
Comment (Vladsz83) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/service/IgniteServiceDeploymentClassLoadingDefaultMarshallerTest.java:168]: Suggestion: a test for re-deploy. Or several. Scenarios: redeploy with empty node filter, redeploy with a known class filter, redeploy with another unknows class ode filter. WDYT?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/LazyServiceConfiguration.java [hunk 8]
```diff
@@ -36,6 +37,10 @@ public class LazyServiceConfiguration extends ServiceConfiguration {
     @GridToStringExclude
     private transient Service srvc;
 
+    /** Node filter. */
+    @GridToStringExclude
+    private transient IgnitePredicate<ClusterNode> nodeFilter;
```
Comment (Vladsz83) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/service/LazyServiceConfiguration.java:42]: `ServiceConfiguration` already has this. Looks wierd.

Comment (Vladsz83) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/service/LazyServiceConfiguration.java:42]: Suggestion: let's emphasize that it is a protection of desacralization failure on base `nodeFilter` until `nodeFilterBytes` is attempted to read.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java [hunk 9]
```diff
@@ -1421,6 +1444,18 @@ private Service copyAndInject(ServiceConfiguration cfg, ServiceContextImpl svcCt
         }
     }
 
+    /** @param cfg Lazy service configuration. */
+    private void unmarshalNodeFilterIfNeeded(LazyServiceConfiguration cfg) throws IgniteCheckedException {
```
Comment (Vladsz83) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java:1448]: Suggestion: method name, some kind of `validateServiceConfiguration`. Which can throw an exception with predefined action description as string. Like 'deploying', 'joining' . Might be useful further, for validation of interceptor class etc.

Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java:1448]: It seems that this method does unmarshalling directly, and should be part of more abstract validation methods - validateNode, etc as it is now

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/LazyServiceConfiguration.java (hunk 8 above)
Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/service/LazyServiceConfiguration.java:42]: It's already emphasized by java key word - transient. Isn't it enough?

Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/service/LazyServiceConfiguration.java:42]: I think comment is much better.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java (hunk 9 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java:1448]: `validateNode` may contain `validateServiceConfig`. Other methods aren't named `validate`

Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java:1448]: But they already do deployment checks, including configuration checks.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java [hunk 10]
```diff
@@ -413,15 +421,26 @@ private void cancelDeployedServices() {
...
+            if (err != null)
+                return new IgniteNodeValidationResult(node.id(), err.getMessage());
+        }
+
+        for (ServiceInfo svc : svcs) {
+            try {
+                unmarshalNodeFilterIfNeeded(svc.configuration());
+            }
+            catch (IgniteCheckedException e) {
+                return new IgniteNodeValidationResult(node.id(), "Node join is rejected: " + e.getMessage());
```
Comment (Vladsz83) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java:441]: A bit info of the node? Like `"Node join is rejected, node=["+node+"]:". Or nodeId at least.

Comment (NSAmelchev) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java]: Done

**Review by Vladsz83 — APPROVED** [review]

**Pull request merged**